File: app/src/main/java/me/zhanghai/android/files/fileproperties/FileLiveData.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.fileproperties;

import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.Path;
import me.zhanghai.android.files.file.FileItem;
import me.zhanghai.android.files.filelist.PathObserver;
import me.zhanghai.android.files.util.CloseableLiveData;

public class FileLiveData extends CloseableLiveData<FileData> {

    @NonNull
    private final Path mPath;

    @NonNull
    private final PathObserver mObserver;

    private volatile boolean mChangedWhileInactive;

    public FileLiveData(@NonNull Path path) {
        this(path, null);
    }

    public FileLiveData(@NonNull FileItem file) {
        this(file.getPath(), file);
    }

    private FileLiveData(@NonNull Path path, @Nullable FileItem file) {
        mPath = path;
        if (file != null) {
            setValue(FileData.ofSuccess(file));
        } else {
            loadValue();
        }
        mObserver = new PathObserver(path, this::onChangeObserved);
    }

    public void loadValue() {
        setValue(FileData.ofLoading());
        AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> {
            FileData value;
            try {
                FileItem file = FileItem.load(mPath);
                value = FileData.ofSuccess(file);
            } catch (Exception e) {
                value = FileData.ofError(e);
            }
            postValue(value);
        });
    }

    private void onChangeObserved() {
        if (hasActiveObservers()) {
            loadValue();
        } else {
            mChangedWhileInactive = true;
        }
    }

    @Override
    protected void onActive() {
        if (mChangedWhileInactive) {
            loadValue();
            mChangedWhileInactive = false;
        }
    }

    @Override
    public void close() {
        mObserver.close();
    }
}

File: app/src/main/java/me/zhanghai/android/files/fileproperties/FileLiveData.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.fileproperties

import android.os.AsyncTask
import java8.nio.file.Path
import me.zhanghai.android.files.file.FileItem
import me.zhanghai.android.files.file.loadFileItem
import me.zhanghai.android.files.util.Failure
import me.zhanghai.android.files.util.Loading
import me.zhanghai.android.files.util.Stateful
import me.zhanghai.android.files.util.Success
import me.zhanghai.android.files.util.valueCompat

class FileLiveData private constructor(
    path: Path,
    file: FileItem?
) : PathObserverLiveData<Stateful<FileItem>>(path) {
    constructor(path: Path) : this(path, null)

    constructor(file: FileItem) : this(file.path, file)

    init {
        if (file != null) {
            value = Success(file)
        } else {
            loadValue()
        }
        observe()
    }

    override fun loadValue() {
        value = Loading(value?.value)
        AsyncTask.THREAD_POOL_EXECUTOR.execute {
            val value = try {
                val file = path.loadFileItem()
                Success(file)
            } catch (e: Exception) {
                Failure(valueCompat.value, e)
            }
            postValue(value)
        }
    }
}
